File: coffee-system/src/main/java/com/jesperancinha/coffee/system/queues/Queue.java
package com.jesperancinha.coffee.system.queues;

import com.jesperancinha.coffee.system.api.queues.CPQueue;
import com.jesperancinha.coffee.system.api.utils.ExecutorServiceHelper;
import lombok.Getter;

import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.Executors;
import java.util.concurrent.ThreadPoolExecutor;

/**
 * Created by joaofilipesabinoesperancinha on 05-05-16.
 */
@Getter
public abstract class Queue implements CPQueue {
    private Map<String, ThreadPoolExecutor> executorServiceMap = new HashMap<>();
    private Map<String, Integer> numberToCreateMap = new HashMap<>();

    /**
     * Number of machines of the same name
     *
     * @param queueSize
     * @param name
     */
    @Override
    public void setQueueSize(int queueSize, String name) {
        Integer currentSize = numberToCreateMap.get(name);
        if (currentSize == null) {
            currentSize = 0;
        }
        numberToCreateMap.put(name, currentSize + queueSize);
    }

    @Override
    public ThreadPoolExecutor getExecutor(String name) {
        return executorServiceMap.get(name);
    }

    @Override
    public void initExecutors() {
        numberToCreateMap.keySet().forEach(
                name -> {
                    final ThreadPoolExecutor currentExecutor = getExecutor(name);
                    if (currentExecutor != null) {
                        ExecutorServiceHelper.shutDownExecutorService(currentExecutor);
                    }
                    final ThreadPoolExecutor managedExecutorService = //
                            (ThreadPoolExecutor) Executors.newFixedThreadPool(numberToCreateMap.get(name));
                    executorServiceMap.put(name, managedExecutorService);
                }
        );
    }

    @Override
    public void stopExecutors() {
        executorServiceMap.values().forEach(
                ExecutorServiceHelper::shutDownExecutorService
        );
    }
}

File: coffee-system/src/main/kotlin/org/jesperancinha/coffee/system/queues/Queue.kt
package org.jesperancinha.coffee.system.queues

import org.jesperancinha.coffee.system.api.utils.ExecutorServiceHelper
import java.util.concurrent.Executors
import java.util.concurrent.ThreadPoolExecutor
import java.util.function.Consumer

/**
 * Created by joaofilipesabinoesperancinha on 05-05-16.
 */
abstract class Queue {
    val executorServiceMap: MutableMap<String, ThreadPoolExecutor> = HashMap()
    val numberToCreateMap: MutableMap<String, Int> = HashMap()

    /**
     * Number of machines of the same name
     *
     * @param queueSize
     * @param name
     */
    fun setQueueSize(queueSize: Int, name: String) {
        var currentSize = numberToCreateMap[name]
        if (currentSize == null) {
            currentSize = 0
        }
        numberToCreateMap[name] = currentSize + queueSize
    }

    fun getExecutor(name: String): ThreadPoolExecutor? {
        return executorServiceMap[name]
    }

    fun initExecutors() {
        numberToCreateMap.keys.forEach(
            Consumer { name: String ->
                val currentExecutor = getExecutor(name)
                if (currentExecutor != null) {
                    ExecutorServiceHelper.shutDownExecutorService(currentExecutor)
                }
                val managedExecutorService =  //
                    Executors.newFixedThreadPool(numberToCreateMap[name]!!) as ThreadPoolExecutor
                executorServiceMap[name] = managedExecutorService
            }
        )
    }

    fun stopExecutors() {
        executorServiceMap.values.forEach { ExecutorServiceHelper.shutDownExecutorService(it) }
    }
}